scipy/sparse/coo.py

```diff
@@ -153,7 +153,7 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
                     M, N = shape
                     self._shape = check_shape((M, N))
 
-                idx_dtype = get_index_dtype(maxval=max(self.shape))
+                idx_dtype = get_index_dtype(maxval=self.shape[0] * self.shape[1])
```

Comment:
This will cause additional memory usage for some users. That may or may not be an issue, but it's worth mentioning.

I may have missed it, but did you try an on-demand approach inside the `reshape` method? That is, detecting the overflow case and promoting to `int64` there?

Reply:
Yes, in the case that both dimensions are less than 2^31 but the maximum linear index is greater than 2^31, this will cause the matrix to require more memory than strictly necessary.

I actually first tried it the way you suggested (upcast only in `reshape`), but it was not a clean solution, so I switched to this solution. The biggest issue that if you upcast at the beginning of `reshape`, then you need to downcast at the end (but only when the dims are still both less than 2^31). If you upcast in the constructor, then you don't need to do anything because `reshape` always preserves the maximum linear index. There is a time-space tradeoff, where upcasting and downcasting can save memory by burning extra cycles on the casting when needed.

I am open to switching back to the upcasting and downcasting technique if saving memory in all cases is paramount.

Reply:
How would you feel about making the constructor less insistent on using the dtype given by `get_index_dtype`? That is, if the user provides index arrays with dtype larger or equal to the minimum dtype, just use them as-is.

```python
minimal_idx_dtype = get_index_dtype(maxval=max(self.shape))
idx_dtype = np.find_common_dtype([row.dtype, col.dtype, minimal_idx_dtype], [])
```

This way we could upcast on demand when reshaping, without having to downcast later.

(Note that my code snippet isn't quite sufficient because `row` and `col` might not be numpy arrays with dtypes, but the idea is the same.)

Reply:
I would find it strange that a matrix with 32-bit indexes becomes a matrix with 64-bit indexes after reshaping, but only sometimes. `reshape` returns `self` if the shape is unchanged, so the upcast would not happen in that case, presumably. It really depends on how predictable you want the dtype behavior to be.

Reply:
Fair point. Let's merge this as-is, then if anyone complains we can explore other options.